### PR TITLE
[python] Implement `tiledbsoma.io.get_experiment_shapes`

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** Analog of `tiledbsoma.io.show_experiment_shapes` for programmatic use as requested on [[sc-64098]](https://app.shortcut.com/tiledb-inc/story/64098/get-output-of-tiledbsoma-io-show-experiment-shapes).

**Changes:**

**Notes for Reviewer:**

Example outputs can be found in the unit-test files.

Stacked atop #3962.

### CODE REVIEW

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) TileDB, Inc. and The Chan Zuckerberg Initiative Foundation
+# Copyright(c) TileDB, Inc.and The Chan Zuckerberg Initiative Foundation
```
[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py:1]: Thanks

And I don't know what tool made these rando formatting mods (I did not)

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 2]
```diff
@@ -302,13 +302,13 @@ def resize_experiment(
...
 
-    # Extra user-provided keys not relevant to the experiment are ignored.  This
+    # Extra user-provided keys not relevant to the experiment are ignored.This
     # is important for the case when a new measurement, which is registered from
     # AnnData/H5AD inputs, is registered and is about to be created but does not
     # exist just yet in the experiment storage.
     #
     # If the user hasn't provided a key -- e.g. a from-anndata-append-with-resize
-    # on one measurement while the experiment's other measurements aren't being
+    # on one measurement while the experiment 's other measurements aren' t being
```
[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py:311]: Yeah IDK :(

[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py:311]: I'll try to do a revert of those mods, and/or re-run the formatter .......

File: apis/python/src/tiledbsoma/io/shaping.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py]: https://github.com/single-cell-data/TileDB-SOMA/pull/3949/commits/53698452d006745f0a0d9a1c68b3cfa0312589d1

File: apis/python/src/tiledbsoma/io/shaping.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py]: https://github.com/single-cell-data/TileDB-SOMA/pull/3949/commits/53698452d006745f0a0d9a1c68b3cfa0312589d1

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 3]
```diff
@@ -718,3 +718,141 @@ def _get_new_ndarray_shape(
...
+
+        return visitor(item, node_name=node_name, args=args, context=context)
+
+
+def get_experiment_shapes(
+    uri: str,
+    *,
+    context: tiledbsoma.SOMATileDBContext | None = None,
+) -> dict[str, Any]:
+    """Similar to ``show_experiment_shapes`` but returns a JSON structure."""
```
[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py]: @jp-dark 
https://github.com/single-cell-data/TileDB-SOMA/pull/3949/commits/af41d9f26c0d2a0c9d2ef8ccdaaeec30ef0c252b

https://tiledbsoma.readthedocs.io/en/kerl-get-experiment-shapes/
is active+hidden (via link only)

**Comment by johnkerl** [comment]
@jp-dark ready for round two! :)

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 4]
```diff
@@ -329,7 +296,13 @@ def _treewalk(
...
     node_name: str | None = None,
     visitor: Callable[..., bool],
-    args: SizingArgs,
+    nobs: int | None,
+    nvars: Dict[str, int] | None,
+    ms_name: str | None,
+    coll_name: str | None,
+    verbose: bool,
+    check_only: bool,
+    output_handle: Printable | None,
```
Comment (jp-dark) [inline apis/python/src/tiledbsoma/io/shaping.py:305]:
```suggestion
    **kwargs
```
The `_treewalk` and `_recurse` methods and the `visitor` call below can/should just forward whatever parameters are passed into the method.

Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py:305]: Thanks! Another commit coming up.

**Comment by johnkerl** [comment]
Moving back to draft while I separate PRs for the reviewer-requested refactor work (`SizingArgs` -> `kwargs`) and the customer-requested UX work (`get_experiment_shapes`), as this (still incomplete) PR is now over 1000 lines and is definitely in need of splitting

**Comment by johnkerl** [comment]
@jp-dark ready for next round of review, with this stacked on top of #3962

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 5]
```diff
@@ -378,9 +468,10 @@ def _recurse(
             retval["var"] = _recurse(item, node_name="var", **kwargs)
             for iter_coll_name in ["X", "obsm", "obsp", "varm", "varp"]:
                 kwargs["coll_name"] = iter_coll_name
-                retval[iter_coll_name] = _recurse(
-                    item, node_name=iter_coll_name, **kwargs
-                )
+                children_dict = _recurse(item, node_name=iter_coll_name, **kwargs)
+                # If the experiment has no varp (etc.) at all, don't include an entry for it.
```
Comment (jp-dark) [inline apis/python/src/tiledbsoma/io/shaping.py:472]: Should we add tests for running on an empty `Experiment` and empty `Measurement`? (Could be a follow-up story for later).

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 6]
```diff
@@ -921,14 +1012,52 @@ def _get_new_ndarray_shape(
...
-    """For vistors which return status dicts, traverses the tree structure
-    and returns True only if all leaf-node statuses are True, else False.
-    """
     if "status" in dikt:
-        value = dikt["status"]
-        assert isinstance(value, bool)
-        return value
+        foo = dikt["status"]
+        assert isinstance(foo, bool)
+        return foo
```
Comment (jp-dark) [inline apis/python/src/tiledbsoma/io/shaping.py:1060]:
```suggestion
        ok = dikt["status"]
        assert isinstance(ok, bool)
        return ok
```

**Review by jp-dark — APPROVED** [review]
Thanks!

File: apis/python/src/tiledbsoma/io/shaping.py (hunk 6 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py:1060]: Oops my bad I didn't mean to leave `foo` in there -- should have proofread more

File: apis/python/src/tiledbsoma/io/shaping.py (hunk 5 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py:472]: Hm, I think that's legit to include in this PR -- 👀

**Comment by johnkerl** [comment]
Dangit I meant to merge #3962 not this one 😠

**Comment by johnkerl** [comment]
I'll make a follow-on for https://github.com/single-cell-data/TileDB-SOMA/pull/3949#discussion_r2040234927

File: apis/python/src/tiledbsoma/io/shaping.py (hunk 5 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/shaping.py:484]: Implemented on #3966
